crates/weaver_cache/src/lib.rs

```diff
@@ -152,65 +185,438 @@ impl Cache {
+
+        // Process the supported formats (i.e.: `.tar.gz`, and `.zip`)
+        if archive_filename.ends_with(TAR_GZ_EXT) {
+            Self::unpack_tar_gz(archive_filename, archive_file, &target_path_buf, sub_folder)?;
+        } else if archive_filename.ends_with(ZIP_EXT) {
+            Self::unpack_zip(archive_filename, archive_file, &target_path_buf, sub_folder)?;
         } else {
-            git_repo_path.to_path_buf()
+            return Err(UnsupportedRegistryArchive {
+                archive: archive_filename.to_owned(),
+            });
         };
 
-        // Adds the repo to the git_repo_dirs hashmap.
-        _ = self
-            .git_repo_dirs
-            .lock()
-            .expect("git_repo_dirs lock failed")
-            .insert(
-                repo_url.clone(),
-                GitRepo {
-                    temp_dir: git_repo_dir,
-                    path: git_repo_path.clone(),
-                },
-            );
-
-        Ok(git_repo_path)
+        Ok(Self {
+            id: id.to_owned(),
+            registry_path,
+            path: target_path_buf,
+            tmp_dir: Some(target_dir),
+        })
+    }
+
+    /// Returns the unique identifier for the registry.
+    #[must_use]
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    /// Unpacks a tar.gz archive into the specified target directory.
+    ///
+    /// This first directory in the archive is skipped as it is the directory corresponding to the
+    /// archive itself. The sub_folder is used to filter the entries to unpack. The sub_folder
+    /// directory is also skipped in the folder hierarchy to only unpack the content of the
+    /// sub-folder.
+    fn unpack_tar_gz(
+        archive_filename: &str,
+        archive_file: File,
+        target_path: &Path,
+        sub_folder: Option<&String>,
+    ) -> Result<(), Error> {
+        let tar_file = flate2::read::GzDecoder::new(archive_file);
+        let mut archive = tar::Archive::new(tar_file);
+
+        for entry in archive.entries().map_err(|e| InvalidRegistryArchive {
+            archive: archive_filename.to_owned(),
+            error: e.to_string(),
+        })? {
+            let mut entry = entry.map_err(|e| InvalidRegistryArchive {
+                archive: archive_filename.to_owned(),
+                error: e.to_string(),
+            })?;
+
+            let path = entry.path().map_err(|e| InvalidRegistryArchive {
+                archive: archive_filename.to_owned(),
+                error: e.to_string(),
+            })?;
+
+            if let Some(valid_entry_path) = Self::path_to_unpack(&path, sub_folder, target_path) {
+                Self::create_parent_dirs(&valid_entry_path, archive_filename)?;
+                // Unpack returns an Unpacked type containing the file descriptor to the
+                // unpacked file. The file descriptor is ignored as we don't have any use for it.
+                _ = entry
```

Comment:
Nit for rust - Maybe we should open a FR - that `#[must_use]` for `Result` where you `?` away a `()` doesn't cause a warning, and then you can `map_error(..)` and something like `.ignore_result()?`

Reply:
I'm not 100% sure, but I think a templated `#[must_use]` function returning a `Result<T, Error>` with `T=()` doesn't emit a warning.

I like the idea of the `.ignore_result()?` if we can specify a parameter `&str` representing the justification.